### PR TITLE
🚧 [WIP] Use pysbd char_span functionality

### PR DESCRIPTION
**Pull request by nipunsadvilkar**
I have been working on this to make it as a spaCy component.

This PR is testing pre-release `pysbd v0.1.5` version having `char_span` functionality.

@danielkingai2 : Use of `char_span` to get spacy `Span` and further setting `is_sent_start` flag on a `Token` would make the codebase more readable and there won't be any need to use custom functions like `merge_segments`.

I have managed to get an abbreviation case right. Rest 4 tests are failing. Would be nice if you have a look at it.

### CODE REVIEW

**Comment by nipunsadvilkar** [comment]
@danielkingai2 , @DeNeutoy Able to pass rest 3 tests. 1 test related to list items is pending. Let me know what you think

File: scispacy/custom_sentence_segmenter.py [hunk 1]
```diff
@@ -31,32 +15,23 @@ def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
...
-    segments = segmenter.segment(doc.text)
-    segments = merge_segments(segments)
-
-    # pysbd splits raw text into sentences, so we have to do our best to align those
-    # segments with spacy tokens
-    segment_index = 0
-    current_segment = segments[segment_index]
-    built_up_sentence = ""
-    for i, token in enumerate(doc):
-        if i == 0 and (token.is_space or token.text == '.'):
```
Comment (danielkingai2) [inline scispacy/custom_sentence_segmenter.py]: Could you add a test for what happens when the document starts with whitespace?

File: scispacy/custom_sentence_segmenter.py [hunk 2]
```diff
@@ -31,32 +15,23 @@ def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
...
-            token.is_sent_start = False
-        elif len(built_up_sentence) >= len(current_segment):
+    segmenter = pysbd.Segmenter(language="en", clean=False, char_span=True)
+    sents_char_spans = segmenter.segment(doc.text)
+
+    char_spans = [
+        doc.char_span(sent_span.start, sent_span.end)
+        for sent_span in sents_char_spans
+    ]
+    start_token_ids = [span[0].idx for span in char_spans if span is not None]
```
Comment (danielkingai2) [inline scispacy/custom_sentence_segmenter.py]: Nit: rename `start_token_ids` to `start_token_char_offsets`

File: scispacy/custom_sentence_segmenter.py [hunk 3]
```diff
@@ -31,32 +15,23 @@ def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
...
-    built_up_sentence = ""
-    for i, token in enumerate(doc):
-        if i == 0 and (token.is_space or token.text == '.'):
-            token.is_sent_start = True
-            continue
-        if token.text.replace('\n', '').replace('\r', '') == '':
-            token.is_sent_start = False
-        elif len(built_up_sentence) >= len(current_segment):
+    segmenter = pysbd.Segmenter(language="en", clean=False, char_span=True)
+    sents_char_spans = segmenter.segment(doc.text)
```
Comment (danielkingai2) [inline scispacy/custom_sentence_segmenter.py]: what is the new return type of `segment`?

File: scispacy/custom_sentence_segmenter.py [hunk 4]
```diff
@@ -31,32 +15,23 @@ def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
...
+    ]
+    start_token_ids = [span[0].idx for span in char_spans if span is not None]
+    for token in doc:
+        prev_token = token.nbor(-1) if token.i != 0 else None
+        if token.idx in start_token_ids:
+            if prev_token and prev_token.text in ABBREVIATIONS:
+                token.is_sent_start = False
+            else:
+                token.is_sent_start = True
+        elif prev_token and prev_token.i != 0 and prev_token.text.count('\n') >= 2:
```
Comment (danielkingai2) [inline scispacy/custom_sentence_segmenter.py]: Will want to adjust the docstring to reflect what you are doing with new lines here.

**Comment by DeNeutoy** [comment]
@nipunsadvilkar are you addressing these fixes upstream in https://github.com/nipunsadvilkar/pySBD/pull/42 ? We'd love to be able to use a stable version of pySBD, it's a great project and has solved a lot of problems for a team internally :)

**Comment by nipunsadvilkar** [comment]
@DeNeutoy Thanks for appreciating. Would it be fine if I fix it over the weekend?

File: requirements.in [hunk 5]
```diff
@@ -20,4 +20,4 @@ mypy
 twine
 
 # required for the tests to run, or to use the custom sentence splitter
-pysbd
+-e git+https://github.com/nipunsadvilkar/pySBD.git@sentence-char-span#egg=pysbd
```
[the remark this answers is not in the document]
Comment (nipunsadvilkar) [reply requirements.in]: switching to pysbd `v0.2.2`

File: scispacy/custom_sentence_segmenter.py (hunk 1 above)
Comment (nipunsadvilkar) [reply scispacy/custom_sentence_segmenter.py]: starting whitespace(s) would be separate sentences.

```python
combined_rule_sentence_segmenter("    This document starts with whitespaces. Next sentence.") 
["    ", "This document starts with whitespaces.", "Next sentence."]
```

This allows us to recreate original text

File: scispacy/custom_sentence_segmenter.py (hunk 3 above)
Comment (nipunsadvilkar) [reply scispacy/custom_sentence_segmenter.py]: `TextSpan` object with (sent_str, start_char, end_char)

File: scispacy/custom_sentence_segmenter.py [hunk 6]
```diff
@@ -31,32 +15,21 @@ def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
...
-            # handle the rare (impossible?) case where spacy tokenizes over a sentence boundary that
-            # pysbd finds
-            built_up_sentence = ' '*int(len(built_up_sentence) - len(current_segment))
-            built_up_sentence = token.text_with_ws
-            segment_index += 1
-            current_segment = segments[segment_index]
+    segmenter = pysbd.Segmenter(language="en", clean=False, char_span=True)
+    sents_char_spans = segmenter.segment(doc.text)
+
+    char_spans = [
```
[the remark this answers is not in the document]
Comment (nipunsadvilkar) [reply scispacy/custom_sentence_segmenter.py]: Yes using pysbd with spacy makes pysbd rely on spacy tokens. spacy requires `tok.is_sent_start` to be set so even if pysbd is able to give proper sentences with its character offsets, it depends on `doc.char_span(sent_span.start, sent_span.end)` to return the Span object. If it returns `None` then we get only one sentence instead of two.

Using only pysbd:
```python
import pysbd 
text = 'How about tomorrow?We can meet at eden garden.' 
seg = pysbd.Segmenter(language="en", clean=False) 
seg.segment(text)
# ['How about tomorrow?', 'We can meet at eden garden.']
```

Using pysbd as a spacy component:
```
doc = nlp(text)
combined_rule_sentence_segmenter(doc)
# ['How about tomorrow?We can meet at eden garden.']

File: scispacy/custom_sentence_segmenter.py (hunk 4 above)
Comment (nipunsadvilkar) [reply scispacy/custom_sentence_segmenter.py]: Again, had to adjust sentencing as per `custom_tokenizer` which handles newlines differently.

Using only pysbd 
```
import pysbd
text = "\n\n2 Long Short-Term Memory Networks\n\n\n\n2.1 Overview\n\nRecurrent neural networks (RNNs) are able to process input sequences of arbitrary leng
th via the recursive application of a transition function on a hidden state vector ht."
seg = pysbd.Segmenter(language="en", clean=False)
```
> ['2 Long Short-Term Memory Networks',
 '2.1 Overview',
 'Recurrent neural networks (RNNs) are able to process input sequences of arbitrary length via the recursive application of a transition function on a hidden state vector ht.']

Comment (nipunsadvilkar) [reply scispacy/custom_sentence_segmenter.py]: & this line does what is already written in docstring.
> New lines that separate sentences will be added to the beginning of the next sentence.

File: scispacy/custom_sentence_segmenter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nipunsadvilkar) [reply scispacy/custom_sentence_segmenter.py]: Adding `xfail` test for it

File: tests/custom_tests/test_custom_segmentation.py [hunk 7]
```diff
@@ -54,20 +52,3 @@ def test_segmenter(en_with_combined_rule_tokenizer_and_segmenter_fixture):
     doc = en_with_combined_rule_tokenizer_and_segmenter_fixture(text)
     # this is really just testing that we handle the case where pysbd crashes
     assert len(list(doc.sents)) > 0
-
```
[the remark this answers is not in the document]
Comment (nipunsadvilkar) [reply tests/custom_tests/test_custom_segmentation.py]: We are no longer using `merge_segments`. What do you suggest to do?

File: scispacy/custom_sentence_segmenter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielkingai2) [reply scispacy/custom_sentence_segmenter.py]: :+1:

File: scispacy/custom_sentence_segmenter.py (hunk 1 above)
Comment (danielkingai2) [reply scispacy/custom_sentence_segmenter.py]: :+1:

File: scispacy/custom_sentence_segmenter.py (hunk 4 above)
Comment (danielkingai2) [reply scispacy/custom_sentence_segmenter.py]: :+1:

File: tests/custom_tests/test_custom_segmentation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielkingai2) [reply tests/custom_tests/test_custom_segmentation.py]: I would suggest adding each of these sentences to the main list of tests, for example, the first of the tests in `test_merge_segment` would end up as this test in the main set of test sentences:
`("This sentence mentions Eqs. 1-4 and should not be split.", ["This sentence mentions Eqs. 1-4 and should not be split."])`. I just want to make sure the behavior that those tests were testing is still handled, even though it is no longer handled by the `merge_segments` function.

File: scispacy/custom_sentence_segmenter.py [hunk 8]
```diff
@@ -5,24 +5,8 @@
...
-
-    # on the off chance an abbreviation ends a sentence and is the end of the document, we just add that segment
-    if temp_segment != "":
-        adjusted_segments.append(temp_segment)
-    return adjusted_segments
+from scispacy.consts import ABBREVIATIONS  # pylint: disable-msg=E0611,E0401
+
 
 def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
     """Adds sentence boundaries to a Doc. Intended to be used as a pipe in a spaCy pipeline.
```
Comment (danielkingai2) [inline scispacy/custom_sentence_segmenter.py]: Could you add to the docstring here to just mention that we leveraging `pysbd` with a link to your repo?

File: scispacy/custom_sentence_segmenter.py [hunk 9]
```diff
@@ -5,24 +5,8 @@
...
-            temp_segment = ""
-
-    # on the off chance an abbreviation ends a sentence and is the end of the document, we just add that segment
-    if temp_segment != "":
-        adjusted_segments.append(temp_segment)
-    return adjusted_segments
+from scispacy.consts import ABBREVIATIONS  # pylint: disable-msg=E0611,E0401
+
 
 def combined_rule_sentence_segmenter(doc: Doc) -> Doc:
```
Comment (danielkingai2) [inline scispacy/custom_sentence_segmenter.py]: The name of this function is left over from a previous iteration, could you rename this to something a bit more descriptive? Like `pysbd_based_sentence_splitter` or `rule_based_sentence_splitter` or `custom_sentence_splitter` or something

**Comment by danielkingai2** [comment]
https://github.com/nipunsadvilkar/pySBD/issues/49 opened an issue for the failing test

**Comment by nipunsadvilkar** [comment]
Yes, I also noticed the same bug yday while fixing the above changes. Will look into it.

**Comment by DeNeutoy** [comment]
yup, lgtm!
